### PR TITLE
GET Linking map of a submission

### PR DESCRIPTION
**Pull request by MightyAx**
ebi-ait/dcp-ingest-central#417

Work In Progress: Using `/submissionEnvelopes/{sub_id}/summary` as a reference.

Census: all object contained within a submission and their relationships to each other.
As @aaclan-ebi's advice, populating links in the direction of process, but not the (theoretically duplicated) opposite direction links from processes to other items.

- [x] Decide on location: `/submissionEnvelopes/{sub_id}/census`
- [x] Decide on document format
- [x] populate data into format
- [x] test speediness
- [x] Do we want to change this to the name `/map`?

### CODE REVIEW

**Comment by MightyAx** [comment]
At this prototype stage which only populates relationship data for processes.
it takes 1 min 30 seconds for core to generate the return file of 3 thousand processes, but transfer time is negligible due to the local network..
Screenshot 2021-11-23 at 10 14 16

**Comment by MightyAx** [comment]
Reversed the mapping direction to make most useful for spreadsheet downloader.
3 Minutes for a complete relationship map of a 3 thousand process submission.
Screenshot 2021-11-23 at 10 58 16

**Comment by aaclan-ebi** [comment]
Sample request and output:
```
$ curl -X GET -H "Accept: application/hal+json" -H "Content-Type: application/hal+json" http://localhost:8080/submissionEnvelopes/619738062807a377aad3a2ff/linkingMap | jq > linking-map.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1073    0  1073    0     0   1407      0 --:--:-- --:--:-- --:--:--  1438
$ cat linking-map.json
{
  "processes": {
    "6197380b2807a377aad3a30e": {
      "protocols": [
        "6197380b2807a377aad3a30a",
        "6197380b2807a377aad3a30b"
      ],
      "inputBiomaterials": [
        "6197380b2807a377aad3a304"
      ],
      "inputFiles": []
    },
    "6197380b2807a377aad3a30d": {
      "protocols": [
        "6197380b2807a377aad3a308",
        "6197380b2807a377aad3a309"
      ],
      "inputBiomaterials": [
        "6197380b2807a377aad3a303"
      ],
      "inputFiles": []
    },
    "6197380b2807a377aad3a30c": {
      "protocols": [
        "6197380b2807a377aad3a307"
      ],
      "inputBiomaterials": [
        "6197380b2807a377aad3a302"
      ],
      "inputFiles": []
    }
  },
  "protocols": {},
  "biomaterials": {
    "6197380b2807a377aad3a304": {
      "derivedByProcesses": [
        "6197380b2807a377aad3a30d"
      ],
      "inputToProcesses": [
        "6197380b2807a377aad3a30e"
      ]
    },
    "6197380b2807a377aad3a303": {
      "derivedByProcesses": [
        "6197380b2807a377aad3a30c"
      ],
      "inputToProcesses": [
        "6197380b2807a377aad3a30d"
      ]
    },
    "6197380b2807a377aad3a302": {
      "derivedByProcesses": [],
      "inputToProcesses": [
        "6197380b2807a377aad3a30c"
      ]
    }
  },
  "files": {
    "6197380b2807a377aad3a306": {
      "derivedByProcesses": [
        "6197380b2807a377aad3a30e"
      ],
      "inputToProcesses": []
    },
    "6197380b2807a377aad3a305": {
      "derivedByProcesses": [
        "6197380b2807a377aad3a30e"
      ],
      "inputToProcesses": []
    }
  }
}
```

**Comment by ke4** [comment]
> Reversed the mapping direction to make most useful for spreadsheet downloader. 3 Minutes for a complete relationship map of a 3 thousand process submission. Screenshot 2021-11-23 at 10 58 16

I can't really decide if it is slow or fast. I would rather say it is extremely slow, but I am not 100% familiar with this feature.

**Review by ke4 — CHANGES_REQUESTED** [review]
I can't see any unit or integration tests. Please add them. Tx

**Review by jacobwindsor — APPROVED** [review]

**Review by ke4 — APPROVED** [review]

**Comment by aaclan-ebi** [comment]
Just wanna note here:  Deployed core in dev returns the linking map in 1min for the same submission https://api.ingest.dev.archive.data.humancellatlas.org/submissionEnvelopes/5cdb0344d96dad000858fff4

```
$ curl -X GET -H "Accept: application/hal+json" -H "Content-Type: application/hal+json" https://api.ingest.dev.archive.data.humancellatlas.org/submissionEnvelopes/5cdb0344d96dad000858fff4/linkingMap | jq > linking-map.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
0     0    0     0    0     0      0      0 --:--:--  0:00:59 --:  0     0    0     0    0     0      0      0 --:--:--  0:00:59 --:--:--     0

````

File: src/main/java/org/humancellatlas/ingest/submission/web/SubmissionLinkMapController.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    @RequestMapping(path = "/submissionEnvelopes/{sub_id}/linkingMap", method = RequestMethod.GET)
+    @ResponseBody
+    public SubmissionLinkingMap getSubmissionLinkMap(@PathVariable("sub_id") SubmissionEnvelope submissionEnvelope) {
+        return new SubmissionLinkingMap(submissionEnvelope);
+    }
+
+    @Getter
+    public class SubmissionLinkingMap {
+        final Dictionary<String, ProcessLinkingMap> processes = new Hashtable<>();
+        final Dictionary<String, UUID> protocols = new Hashtable<>();
```
Comment (MightyAx) [inline src/main/java/org/humancellatlas/ingest/submission/web/SubmissionLinkMapController.java:39]: I think this line is unused. Protocols are now returned as part of `this.processes`

**Pull request merged**